### PR TITLE
[FLINK-32489][runtime] Support serialize and merge BloomFilter for runtime filter

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

*Runtime filter needs to use BloomFIlter, it would be transferred from runtime filter builder operator to runtime filter operator via network, so serialization and merge need.*

## Brief change log

  - *Support serialize and merge BloomFilter for runtime filter*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit test in BloomFIlterTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no)
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by lsyldliu** [comment]
cc @wanglijie95

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BitSet.java [hunk 1]
```diff
@@ -83,6 +83,13 @@ public int bitSize() {
         return bitLength;
     }
 
+    public byte[] getData() {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BitSet.java:86]: Maybe `getData` -> `toBytes` or `toByteArray`, and add doc to indicate that currently only supports heap.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BitSet.java [hunk 2]
```diff
@@ -83,6 +83,13 @@ public int bitSize() {
...
+        if (memorySegment.isOffHeap()) {
+            throw new IllegalStateException("Only support use heap memory for serialization");
+        }
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BitSet.java:89]: This `if` block can repalced by the `checkState`

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java [hunk 3]
```diff
@@ -135,6 +144,30 @@ public void reset() {
         this.bitSet.clear();
     }
 
+    /**
+     * Merge the specified bloom filter with current bloom filter.
+     *
+     * @param that - bloom filter to merge
+     */
+    public void merge(BloomFilter that) {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:152]: Is this method needed? I think we only need the `mergeBloomFilterBytes`

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java [hunk 4]
```diff
@@ -143,4 +176,63 @@ public String toString() {
         output.append(bitSet);
         return output.toString();
     }
+
+    public static byte[] toBytes(BloomFilter filter) {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:180]: Add doc, and indicate that currently only supports heap.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java [hunk 5]
```diff
@@ -143,4 +176,63 @@ public String toString() {
...
+        int byteSize = data.length;
+        byte[] bytes = new byte[8 + byteSize];
+        UNSAFE.putInt(bytes, BYTE_ARRAY_BASE_OFFSET, filter.numHashFunctions);
+        UNSAFE.putInt(bytes, BYTE_ARRAY_BASE_OFFSET + 4, byteSize);
+        UNSAFE.copyMemory(
+                data, BYTE_ARRAY_BASE_OFFSET, bytes, BYTE_ARRAY_BASE_OFFSET + 8, byteSize);
+        return bytes;
+    }
+
+    public static BloomFilter fromBytes(byte[] bytes) {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:191]: Add doc, and indicate that currently only supports heap.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java [hunk 6]
```diff
@@ -43,17 +45,24 @@
...
 
     public BloomFilter(int expectedEntries, int byteSize) {
         checkArgument(expectedEntries > 0, "expectedEntries should be > 0");
-        this.expectedEntries = expectedEntries;
         this.numHashFunctions = optimalNumOfHashFunctions(expectedEntries, byteSize << 3);
         this.bitSet = new BitSet(byteSize);
     }
 
+    /** A constructor to support rebuilding the BloomFilter from a serialized representation. */
+    private BloomFilter(BitSet bitSet, int numHashFunctions) {
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:61]: check `bitSet` not null, and check `numHashFunctions` > 0

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java [hunk 7]
```diff
@@ -143,4 +176,63 @@ public String toString() {
...
+        bitSet.setMemorySegment(MemorySegmentFactory.wrap(data), 0);
+        return new BloomFilter(bitSet, numHashFunctions);
+    }
+
+    public static byte[] mergeBloomFilterBytes(byte[] bf1Bytes, byte[] bf2Bytes) {
+        return mergeBloomFilterBytes(bf1Bytes, 0, bf1Bytes.length, bf2Bytes, 0, bf2Bytes.length);
+    }
+
+    /** Merge the bf2 bytes to bf1 */
+    public static byte[] mergeBloomFilterBytes(
```
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:208]: can be private

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java [hunk 8]
```diff
@@ -168,4 +173,161 @@ public void testHashcodeInput() {
         assertTrue(bloomFilter.testHash(val4));
         assertTrue(bloomFilter.testHash(val5));
     }
+
+    @Test
+    public void testBloomFilterSerDe() {
```
Comment (wanglijie95) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java:178]: testBloomFilterSerializationAndDeserialization

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java (hunk 7 above)
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/operators/util/BloomFilter.java:208]: Maybe named `mergeSerializedBloomFilters` ? And add doc to indicate that `bf1Bytes` will be changed after the merge.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java (hunk 8 above)
Comment (wanglijie95) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java:178]: Can we migrate this test to junit5?

**Review by wanglijie95 — COMMENTED** [review]
@lsyldliu Thanks for creating this PR. I left some comments, PTAL.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java (hunk 8 above)
Comment (lsyldliu) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/BloomFilterTest.java:206]: Yes, I have migrated all tests to junit5

**Comment by lsyldliu** [comment]
@wanglijie95 Thanks for your review, I've updated it.

**Review by wanglijie95 — APPROVED** [review]
Thanks @lsyldliu . LGTM now.

**Pull request closed without merging**